Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIActionCocoa.mm

```diff
@@ -0,0 +1,305 @@
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionContext.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "CocoaHelpers.h"
+#import "WebExtensionAction.h"
+#import "WebExtensionContextProxyMessages.h"
+#import "WebExtensionUtilities.h"
+
+namespace WebKit {
+
+static RefPtr<WebExtensionAction> getActionWithIdentifiers(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, WebExtensionContext& context, NSString *apiName, std::optional<String>& error)
+{
+    RefPtr<WebExtensionAction> action;
+
+    if (windowIdentifier) {
+        auto window = context.getWindow(windowIdentifier.value());
+        if (!window) {
+            error = toErrorString(apiName, nil, @"window not found");
+            return nullptr;
+        }
+
+        return context.getAction(window.get()).ptr();
+    }
+
+    if (tabIdentifier) {
+        auto tab = context.getTab(tabIdentifier.value());
+        if (!tab) {
+            error = toErrorString(apiName, nil, @"tab not found");
+            return nullptr;
+        }
+
+        return context.getAction(tab.get()).ptr();
+    }
+
+    return &context.defaultAction();
+}
+
+static RefPtr<WebExtensionAction> getOrCreateActionWithIdentifiers(std::optional<WebExtensionWindowIdentifier> windowIdentifier, std::optional<WebExtensionTabIdentifier> tabIdentifier, WebExtensionContext& context, NSString *apiName, std::optional<String>& error)
+{
+    RefPtr<WebExtensionAction> action;
+
+    if (windowIdentifier) {
+        auto window = context.getWindow(windowIdentifier.value());
+        if (!window) {
+            error = toErrorString(apiName, nil, @"window not found");
+            return nullptr;
+        }
+
+        return context.getOrCreateAction(window.get()).ptr();
+    }
+
+    if (tabIdentifier) {
+        auto tab = context.getTab(tabIdentifier.value());
+        if (!tab) {
+            error = toErrorString(apiName, nil, @"tab not found");
+            return nullptr;
+        }
+
+        return context.getOrCreateAction(tab.get()).ptr();
+    }
+
+    return &context.defaultAction();
```

Comment:
Does this do a `getOrCreate` under the hood?

Reply:
Yeah, created on demand.

```
WebExtensionAction& WebExtensionContext::defaultAction()
{
    if (!m_defaultAction)
        m_defaultAction = WebExtensionAction::create(*this);

    return *m_defaultAction;
}
```